selfdrive/locationd/magd.py

```diff
@@ -0,0 +1,221 @@
+          cache_CP = msg
+        if self.get_restore_key(cache_CP, cache_mc.version) == self.get_restore_key(CP, VERSION):
+          if cache_mc.calibrated:
+            self.calibrationParams = np.array(cache_mc.calibrationParams)
+          self.point_buckets.load_points(cache_mc.points)
+          cloudlog.info("restored magnetometer calibration params from cache")
+      except Exception:
+        cloudlog.exception("failed to restore cached magnetometer calibration params")
+        params.remove("MagnetometerCarParams")
+        params.remove("MagnetometerCalibration")
+
+  def get_restore_key(self, CP: car.CarParams, version: int) -> Tuple[str, int]:
+    return (CP.carFingerprint, version)
+
+  def reset(self) -> None:
+    self.calibrationParams = CALIB_DEFAULTS
+    self.calibrated = False
+    self.bearingValid = False
+    self.filtered_magnetic = [FirstOrderFilter(0, FILTER_DECAY, FILTER_DT) for _ in range(3)]
+    self.filtered_vals = np.zeros(3)
+    self.vego = 0
+    self.yaw = 0
+    self.yaw_valid = False
+    self.point_buckets = MagBuckets(x_bounds=BUCKET_KEYS,
+                                        min_points=[MIN_BUCKET_POINTS] * len(BUCKET_KEYS),
+                                        min_points_total=MIN_POINTS_TOTAL,
+                                        points_per_bucket=POINTS_PER_BUCKET,
+                                        rowsize=3)
+    self.past_raw_vals = np.zeros(3)
+
+  def get_ellipsoid_rotation(self, coeffs: np.ndarray) -> Tuple[np.ndarray, float]:
+    a, b, c = coeffs[:3]
+    M = np.array([
+        [2 * a, c],
+        [c, 2 * b]
+    ])
+    eigenvalues, eigenvectors = np.linalg.eigh(M)
+    axes_lengths = np.sqrt(1 / eigenvalues)
+    rotation_angle = np.arctan2(eigenvectors[1, 0], eigenvectors[0, 0])
+    return axes_lengths, rotation_angle
+
+  def get_ellipsoid_center(self, coeffs: np.ndarray) -> np.ndarray:
+    a, b, c, d, e = coeffs[:5]
+    num = (c**2 / 4) - (a * b)
+    x0 = ((b * d / 2) - (c / 2 * e / 2)) / num
+    y0 = ((a * e / 2) - (c / 2 * d / 2)) / num
+    return np.array([x0, y0])
+
+  def estimate_calibration_params(self) -> np.ndarray:
+    points = self.point_buckets.get_points(FIT_POINTS_TOTAL)
+    calibration_params = np.ones(5) * np.nan
+    try:
+      x, y, yaw = points[:, 0], points[:, 1], points[:, 2]
+      D = np.vstack((x * x, y * y, x * y, x, y, np.ones_like(x))).T
+      _, _, V = np.linalg.svd(D, full_matrices=False)
+      coeffs = V[-1] / V[-1, -1]
+      x0, y0 = self.get_ellipsoid_center(coeffs)
+      (l1, l2), theta = self.get_ellipsoid_rotation(coeffs)
+      bearing = self.get_calibrated_bearing(x, y, np.array([x0, y0, l1, l2, theta, 0.0]))
+      offset = np.median(self.reset_angle_range(bearing - yaw))
+      calibration_params = np.array([x0, y0, l1, l2, theta, offset])
+    except np.linalg.LinAlgError as e:
+      cloudlog.exception(f"LinAlgError computing magnetometer calibration params: {e}")
```

Comment:
don't need the `e`, cloudlog.exception will log the exception nicely
```suggestion
      cloudlog.exception("LinAlgError computing magnetometer calibration params")
```